common/featureset.cpp

```diff
@@ -0,0 +1,55 @@
+#include "featureset.h"
+#include <QMap>
+#include <QDebug>
+
+FeatureSet::FeatureSet()
+{
+
+}
+
+QMap<QString, bool> FeatureSet::getDefaultFeatureList() {
+    initalizeFeatureList(featureList);
+    return featureList;
+}
+
+void FeatureSet::initalizeFeatureList(QMap<QString, bool> &featureList){
+    featureList.insert("client_id", false);
+    featureList.insert("client_ver", false);
+    featureList.insert("feature_set", false);
+}
+
+void FeatureSet::enableRequiredFeature(QMap<QString, bool> &featureList, QString featureName){
+    if (featureList.contains(featureName))
+        featureList.insert(featureName,true);
+}
+
+void FeatureSet::disableRequiredFeature(QMap<QString, bool> &featureList, QString featureName){
+    if (featureList.contains(featureName))
+        featureList.insert(featureName,false);
+}
+
+QMap<QString, bool> FeatureSet::addFeature(QMap<QString, bool> &featureList, QString featureName, bool isFeatureRequired){
+    featureList.insert(featureName,isFeatureRequired);
+    return featureList;
+}
+
+QMap<QString, bool> FeatureSet::identifyMissingFeatures(QMap<QString, bool> suppliedFeatures, QMap<QString, bool> requiredFeatures){
+    QMap<QString, bool> missingList;
+    QMap<QString, bool>::iterator i;
+    for (i = requiredFeatures.begin(); i != requiredFeatures.end(); ++i) {
+        if (!suppliedFeatures.contains(i.key())) {
+            missingList.insert(i.key(), i.value());
+        }
+    }
+    return missingList;
+}
```

Comment:
(continuing previous discussion here)

I think what you have works, I was just looking for a way to make it clearer. I don't think C++/Qt gives us the tools to make it easy.

In scala I'd do something like...
```scala
val requiredFeatures: Set[String] = serverFeatures.filter { case (k, v) => v }.keySet
val missingFeatures: Set[String] = requireFeatures - clientFeatures // Set difference
if (missingFeatures.nonEmpty) ...
```

Doing it with `any` (`exists` in scala)
```scala
clientFeatures: Set[String] = ???
if (clientFeatures.exists(f => serverFeatures.getOrElse(f, false)) ...
```

The latter might translate (haven't tried to compile something like this)
```C++
QSet<String> clientFeatures;
if (std::any_of(clientFeatures.begin(), clientFeatures.end(), [](??? featureName) { serverFeatures.value(featureName, false) })) { ... }
```